### PR TITLE
Full site editor phase one

### PR DESCRIPTION
**Pull request by Camwyn**
[TEC-4262]

Adds a TCMN `tec_is_full_site_editor` function to use as a check for FSE themes.

Also tweaks some asset and implementation loading  to handle requirements.

In case you are wondering, TBH I am not jazzed about the namespacing. Open to suggestions.

Required by https://github.com/the-events-calendar/the-events-calendar/pull/3779

[TEC-4262]: https://theeventscalendar.atlassian.net/browse/TEC-4262?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Common/Editor/Full_Site/Hooks.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+namespace TEC\Common\Editor\Full_Site;
+
+/**
+ * Class Hooks
+ *
+ * @since   TBD
+ *
+ * @package TEC\Common\Editor\Full_Site
+ */
+class Hooks extends \tad_DI52_ServiceProvider {
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Hooks.php:12]: The `register` method of this provider is empty: is this required?

File: src/Common/Editor/Full_Site/Assets.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class Assets extends \tad_DI52_ServiceProvider {
+
+	/**
+	 * Binds and sets up implementations.
+	 *
+    * @since TBD
+	 */
+	public function register() {
+		$plugin = Common::instance();
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Assets.php:22]: Is this required and not done as part of `Main`? If not required, then the whole service provider should be dropped.

File: src/Common/Editor/Full_Site/Provider.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		$this->container->singleton( static::class, $this );
+	}
+
+	/**
+	 * Registers the provider handling all the assets for this Service Provider
+	 *
+	 * @since TBD
+	 */
+	protected function register_assets() {
+		$assets = new Assets( $this->container );
```
[the remark this answers is not in the document]
Comment (lucatume) [reply src/Common/Editor/Full_Site/Provider.php:40]: Additionally, to register a sub-service provider from a provider, you can just use `$this->container->register(Provider::class)`.

File: src/Common/Editor/Full_Site/Provider.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		$this->container->singleton( Assets::class, $assets );
+	}
+
+	/**
+	 * Registers the provider handling all the 1st level filters and actions for this Service Provider.
+	 *
+	 * @since TBD
+	 */
+	protected function register_hooks() {
+		$hooks = new Hooks( $this->container );
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Provider.php:52]: Same note about registering sub-service providers here.

File: src/Common/Editor/Full_Site/Template_Utils.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * @package TEC\Common\Editor\Full_Site
+ */
+class Template_Utils {
+	/**
+	 * Returns an array containing the references of the passed blocks and their inner blocks.
+	 *
+	 * @since TBD
+	 *
+	 * @param array $blocks array of blocks.
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Template_Utils.php:18]: Add the array types here?

File: src/Common/Editor/Full_Site/Template_Utils.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	 *
+	 * @return array block references to the passed blocks and their inner blocks.
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Template_Utils.php:20]: Add the array types here?

File: src/Common/Editor/Full_Site/Template_Utils.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+			$queue_count = count( $queue );
+		}
+
+		return $all_blocks;
+	}
+
+	/**
+	 * Parses wp_template content and injects the current theme's stylesheet as a theme attribute into
+	 * each wp_template_part
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Template_Utils.php:49]:
```suggestion
	 * each wp_template_part.
```

File: src/Tribe/Main.php [hunk 8]
```diff
@@ -155,6 +156,16 @@ public function context_class() {
 		return $this->plugin_context_class;
 	}
 
+	/**
+	 * Initializes all libraries used/required by our singletons.
+
```
Comment (lucatume) [inline src/Tribe/Main.php:161]:
```suggestion
```

File: src/Tribe/Main.php [hunk 9]
```diff
@@ -155,6 +156,16 @@ public function context_class() {
...
+	 *
+	 * @since TBD
+	 */
+	public function init_early_libraries(): void {
```
Comment (lucatume) [inline src/Tribe/Main.php:165]:
```suggestion
	public function init_early_libraries() {
```
Check for others PHP 5.6 incompatibilities.

File: src/Tribe/Main.php [hunk 10]
```diff
@@ -671,6 +682,8 @@ public function bind_implementations() {
 		tribe_register_provider( Tribe\Service_Providers\Onboarding::class );
 		tribe_register_provider( Tribe\Admin\Notice\Service_Provider::class );
 		tribe_register_provider( Tribe\Admin\Conditional_Content\Service_Provider::class );
+
```
Comment (lucatume) [inline src/Tribe/Main.php:685]:
```suggestion
```
The blank line in between lends gravitas to the registration of the full-site editor provider, intended?

File: src/functions/editor.php [hunk 11]
```diff
@@ -0,0 +1,22 @@
+<?php
+/**
+ * Check if the current theme is a block theme.
+ *
+ * @since TBD
+ *
+ * @return bool
```
Comment (lucatume) [inline src/functions/editor.php:7]:
```suggestion
 * @return bool Whether the current theme is a block theme supporting full-site editing.
```

**Review by lucatume — CHANGES_REQUESTED** [review]

File: src/Common/Editor/Full_Site/Hooks.php (hunk 1 above)
Comment (Camwyn) [reply src/Common/Editor/Full_Site/Hooks.php:12]: Another one that will be used in the next iteration.

File: src/Common/Editor/Full_Site/Provider.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Camwyn) [reply src/Common/Editor/Full_Site/Provider.php:40]: Actually - with the removal of Assets and Hooks, we don't need the service provider at all - so I'll remove that too (for now, until we actually need it)

File: src/Common/Editor/Full_Site/Template_Utils.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+	/**
+	 * Returns an array containing the references of the passed blocks and their inner blocks.
+	 *
+	 * @since TBD
+	 *
+	 * @param array $blocks array of blocks.
+	 *
+	 * @return array block references to the passed blocks and their inner blocks.
+	 */
+	public static function flatten_blocks( &$blocks ) {
```
[the remark this answers is not in the document]
Comment (Camwyn) [reply src/Common/Editor/Full_Site/Template_Utils.php:22]: 1. When we return we are replacing/overwriting `$blocks` with `$all_blocks` - and that's what the pass-by-reference is for.
2. If we don't pass it by reference the non-event blocks get lost (header and footer)

File: package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "tribe-common",
-  "version": "4.14.19",
+  "version": "4.14.18",
```
Comment (borkweb) [inline package.json:3]: This change smells fishy.

File: src/Common/Editor/Full_Site/Template_Utils.php [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+ */
+class Template_Utils {
+	/**
+	 * Returns an array containing the references of the passed blocks and their inner blocks.
+	 *
+	 * @since TBD
+	 *
+	 * @return array<array<string,mixed>> Array of parsed block objects.
+	 *
+	 * @return array<array<string,mixed>> Block references to the passed blocks and their inner blocks.
```
Comment (borkweb) [inline src/Common/Editor/Full_Site/Template_Utils.php:20]: Pick one :D

**Review by borkweb — CHANGES_REQUESTED** [review]

File: package.json (hunk 13 above)
Comment (Camwyn) [reply package.json:3]: 4.14.18 was macaron - it got nixed, so I had to back up the version #
See: https://docs.google.com/spreadsheets/d/1m5xmULGTpqRAQzMG-pVT74Ub9AnSsOT87yGFz7toLTE/edit#gid=1365577016
`4.14.18` is correct

Comment (Camwyn) [reply package.json:3]: Now I need to go check mooncake's version numbers

File: src/Common/Editor/Full_Site/Template_Utils.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (lucatume) [reply src/Common/Editor/Full_Site/Template_Utils.php:22]: Ok, I can see how the same question might come up again, I would add your answer to the long description section of the doc-block.

File: src/functions/editor.php [hunk 15]
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * Check if the current theme is a block theme.
+ *
+ * @since TBD
+ *
+ * @return bool Whether the current theme is a block theme supporting full-site editing.
+ */
+if ( ! function_exists( 'tec_is_full_site_editor' ) ) {
+	function tec_is_full_site_editor() {
```
Comment (lucatume) [inline src/functions/editor.php:10]: To be correctly parsed by the documentation generators and IDEs, the doc-block should be moved before the definition of the `tec_is_sull_site_editor` function, not before the check to see if it's defined already.

**Review by lucatume — CHANGES_REQUESTED** [review]

File: src/Tribe/Main.php (hunk 10 above)
Comment (Camwyn) [reply src/Tribe/Main.php]: No, and I love that you used "gravitas"

File: src/Common/Editor/Full_Site/Template_Utils.php [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @since   TBD
+ *
+ * @package TEC\Common\Editor\Full_Site
+ */
+class Template_Utils {
+	/**
+	 * Returns an array containing the references of the passed blocks and their inner blocks.
+	 *
+     * When we return we are replacing/overwriting $blocks with $all_blocks so we pass-by-reference.
```
Comment (lucatume) [inline src/Common/Editor/Full_Site/Template_Utils.php:16]: Indentation is off here.

**Review by lucatume — CHANGES_REQUESTED** [review]

**Review by lucatume — APPROVED** [review]

**Pull request merged**